### PR TITLE
Add --insecure to pull, push and login (carrying 10719)

### LINKED ISSUE #8887: --insecure-registry should be on "docker pull"
Hi folks, thanks for all your great work.

I was previously running a "library/registry" on `localhost:5000`. With Docker 1.3+, I was required to run *docker* with `--insecure-registry localhost:5000`.  Doing so did nothing, until I discovered I needed to run **docker**, as in **daemon**, with those parameters.

It would be very useful to have that handled directly by `docker pull`, and not have to restart the whole thing and tweak system-level settings when you discover you need to use a local unsecure registry.  **EDIT:** As mentioned in the comments, it would also be very useful to allow *any* registry to be unsecure, not just named ones, as Docker sometimes provides random ports, and some environments have many registries popping in and out of existence.

It is currently read here: https://github.com/docker/docker/blob/master/docker/daemon.go#L43 (while running the daemon), and it's checked while `pull`ing in https://github.com/docker/docker/blob/master/graph/pull.go#L116 .. maybe we could add yet another switch to `pull` like `--insecure` and tweak that would forcefully make it `secure == false` ?

I don't have a docker development setup ready, but if you think it's a good idea.. I could try to implement it.

```
Linux cerise 3.13.0-32-generic #57-Ubuntu SMP Tue Jul 15 03:51:08 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux
Docker version 1.3.1, build 4e9bbfa
Containers: 5
Images: 607
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Dirs: 618
Execution Driver: native-0.2
Kernel Version: 3.13.0-32-generic
Operating System: Ubuntu 14.04.1 LTS
Debug mode (server): false
Debug mode (client): true
Fds: 10
Goroutines: 11
EventsListeners: 0
Init Path: /usr/bin/docker
Username: abourget
Registry: [https://index.docker.io/v1/]
WARNING: No swap limit support
```

### PR DESCRIPTION
**Pull request by tiborvass**
Fixes #8887

Add `--insecure` to pull, push and login.

I carried @jsdir's work.

### CODE REVIEW

**Comment by jfrazelle** [comment]
I'm not really comfortable with --insecure on login

**Comment by tiborvass** [comment]
Ping @ewindisch @NathanMcCauley

**Comment by dmp42** [comment]
cc @docker/distribution-maintainers

**Comment by endophage** [comment]
I agree with @jfrazelle re login. If somebody isn't willing to set up TLS to secure users' passwords on the network, they shouldn't be doing authentication.

**Comment by Savar** [comment]
I disagree with @endophage and @jfrazelle. You try to enforce everybody to do it as you think it is right.

1. in terms of "MITM".. keep in mind that it seems to be quiet easy to get a valid certificate for somebody else (at least if you are one of the bad guy with enough money)

2. if you pull an image and runs it you're probably running in a "root" context which might be able to break out.. so it is almost much more worse to run an image which might be compromised than losing your credentials..

So in my opinion this topic was discussed a lot (I can remember 5 or 6 pull requests) and you still try to fight against this option which is in my case eyewashing.

Better add an option to enforce a fingerprint for a certificate.. this would even make it much easier to use TLS with a self signed certificate because everybody could specify it on a pull/push/login cmd...

**Comment by diogomonica** [comment]
@Savar:

As far as your argument 1) goes, that is not true. There actually have been surprisingly few incidents over the past few years where attackers have been able to get in possession of certificates for domains they didn't own. If this was true, then no one should be doing online banking using browsers and TLS.

2) Is also not a good argument. There is a significant difference between an active MiTM and a passive sniffing attack. It is a lot harder for an attacker to setup a malicious Docker registry that intercepts legitimate requests and serves malicious images, than to simple sniff the credentials out of the air/wire.

Supporting --insecure on login is the wrong call.

**Comment by Savar** [comment]
@diogomonica

2) it might be harder, but it would be not that hard to even wrap the container you're pulling from the original repository.. so this is not a good argument against my argument but I agree that it is much easier to upload "bad" images afterwards if you get the credentials...

but the RISK is more or less the same.. if I know about docker and I want to do something bad, even if I have no credentials I will hook up and send you a bad image.. maybe automatically wrapping the original image you wanted to pull..

so if you have --insecure on pull .. you shoulnd't care about --insecure on login..

3) if you have your private network which is fully in your control (company network..) it might be impossible to do a MITM but you still want to use TLS.. (even if it is then not really necessary/helpful)

4) the initial idea to force everybody to install the CA certificate into your system is also not 100% the right way.. in that case the company (or a bad admin) can really do the MITM because now the company can be google/microsoft/....

At the end it would be nicer if docker would have something like a local cert storage (like Firefox/Chrome/...) where you can install your certificate easily.. that would probably solve most of the issues.

**Comment by dmcgowan** [comment]
> so if you have --insecure on pull .. you shoulnd't care about --insecure on login..

As soon as an --insecure login is allowed, now every --insecure pull will be transmitting credentials in the plaintext, this is what we are actively avoiding and something you should care about. If you don't care, then don't require credentials on the registry.

We are actively designing to improve this use case for the v2 registry, mostly involving easier ways to setup trust between a daemon and registry and avoid sending credentials every time. However we should not support making the existing use case less secure in the meantime. If you are passionate about trust/distribution related topics, I encourage you to head over the distribution project and get involved there https://github.com/docker/distribution.

**Comment by tiborvass** [comment]
@diogomonica @dmcgowan I'm fine removing `login --insecure`, would it be accepted if I did only pull --insecure and push --insecure ?

**Comment by dmcgowan** [comment]
I would prefer a check that the index is not official.  I am not fan of additional special checks but it shouldn't be possible to overwrite a known secure endpoint on the client end.

**Comment by endophage** [comment]
I'm ok with pull and push supporting --insecure, but I also agree with @dmcgowan. It would be better if it indicates that an insecure registry is acceptable but should not force a fallback if the registry supports TLS.

**Comment by dmcgowan** [comment]
It will not force a fallback, the insecure flag will still prefer HTTPS, only allow fallback to HTTP

File: docs/man/docker-pull.1.md [hunk 1]
```diff
@@ -19,7 +19,9 @@ It is also possible to specify a non-default registry to pull from.
 
 # OPTIONS
 **-a**, **--all-tags**=*true*|*false*
-   Download all tagged images in the repository. The default is *false*.
+  Download all tagged images in the repository. The default is *false*.
+**-i**, **--insecure-registry**=*true*|*false*
```
Comment (stevvooe) [inline docs/man/docker-pull.1.md]: There really shouldn't be a short flag for insecure. If one wants this, please type it out.

File: api/client/commands.go [hunk 2]
```diff
@@ -1289,6 +1296,7 @@ func (cli *DockerCli) CmdImport(args ...string) error {
 
 func (cli *DockerCli) CmdPush(args ...string) error {
 	cmd := cli.Subcmd("push", "NAME[:TAG]", "Push an image or a repository to the registry", true)
+	insecure := cmd.Bool([]string{"-insecure"}, false, "Allow communication with insecure registries")
```
Comment (stevvooe) [inline api/client/commands.go]: This doesn't match documentation. Should this be "-insecure-registry" or should the docs be change to "--insecure"?

**Comment by stevvooe** [comment]
@dmcgowan I'm wondering if the flag should be something like "--allow-http-fallback" or something more descriptive of what is actually happening.

**Comment by dmcgowan** [comment]
The `--insecure` flag matches the daemon side flag of `--insecure-registries`.  Although it is just used for http fallback, it will likely be used in the future as an escape hatch for installing from untrusted registries as well (possibly any v1 registry).  If you really want the flag to be more descriptive of what is happening, might I suggest `--i-dont-want-to-properly-configure-my-daemon-or-registry`.

**Comment by tiborvass** [comment]
@stevvooe @dmcgowan it's not only used for http fallback, it's also used for SkipVerify for HTTPS.

**Comment by xiaods** [comment]
not comfortable with this feature. do we indeed preder this flag --insecure?

**Comment by diogomonica** [comment]
My general feeling is that we should be pushing for a safer Docker ecosystem, and this "feature" being merged gets us further from this goal.

**Comment by ewindisch** [comment]
It's also worth noting that there are a variety of workarounds for this.

I'm not terribly comfortable with this, but it's also of the operator's choice to do something terrible and/or shoot themselves in the foot. Such discussions make me question if the engine should have a tainted flag as in the Linux kernel.

**Comment by tiborvass** [comment]
@ewindisch @diogomonica @NathanMcCauley if one of you could formulate a final decision on this, that would be very helpful. Make sure to take into account all the users' needs that were +1-ing all those closed PRs / issues. In the meantime, I'm closing this since it cannot be merged.

**Pull request closed without merging**